Microsoft.WindowsAzure.Storage/src/timer_handler.cpp

```diff
@@ -41,67 +40,66 @@ namespace azure {    namespace storage {    namespace core {
             m_cancellation_token.deregister_callback(m_cancellation_token_registration);
         }
 
-#ifdef _WIN32
         stop_timer();
-#else // LINUX
-        try
-        {
-            stop_timer();
-        }
-        catch (boost::exception_detail::clone_impl<boost::exception_detail::error_info_injector<boost::system::system_error> > &e)
-        {
-        }
-#endif
-
     }
 
     void timer_handler::start_timer(const std::chrono::milliseconds& time)
     {
-        m_mutex = std::make_shared<std::mutex>();
-        auto this_pointer = std::dynamic_pointer_cast<timer_handler>(shared_from_this());
-        m_timeout_task = timeout_after(time).then([this_pointer]()
+        std::lock_guard<std::mutex> guard(m_mutex);
+        if (m_timer_started.load(std::memory_order_acquire))
+        {
+            return;
+        }
+        m_timer_started.store(true, std::memory_order_release);
+        std::weak_ptr<timer_handler> weak_this_pointer = shared_from_this();
+        m_timeout_task = timeout_after(time).then([weak_this_pointer]()
         {
-            this_pointer->m_is_canceled_by_timeout = true;
-            this_pointer->m_worker_cancellation_token_source->cancel();
+            auto this_pointer = weak_this_pointer.lock();
+            if (this_pointer)
+            {
+                this_pointer->m_is_canceled_by_timeout.store(true, std::memory_order_release);
+                this_pointer->m_worker_cancellation_token_source.cancel();
+            }
         });
     }
 
     void timer_handler::stop_timer()
     {
-        if (m_timer != nullptr)
+        std::lock_guard<std::mutex> guard(m_mutex);
+        if (m_timer_started.load(std::memory_order_acquire) && m_timer)
```

Comment:
>m_timer_started [](start = 12, length = 15)

Should it be set to `false` after timer has been stopped?

Reply:
I think the timer isn't supposed to start again after it stops. So we need this bool to avoid starting twice.